File: intrinsic/logging/utils/downsampler/proto_conversion_test.cc
// Copyright 2023 Intrinsic Innovation LLC

#include "intrinsic/logging/utils/downsampler/proto_conversion.h"

#include <gmock/gmock.h>
#include <gtest/gtest.h>

#include <optional>

#include "absl/status/status.h"
#include "absl/time/time.h"
#include "intrinsic/logging/proto/downsampler.pb.h"
#include "intrinsic/logging/utils/downsampler/downsampler.h"
#include "intrinsic/util/testing/gtest_wrapper.h"

namespace intrinsic {
namespace {

using ::intrinsic::data_logger::DownsamplerEventSourceState;
using ::intrinsic::data_logger::DownsamplerOptions;
using ::intrinsic::data_logger::DownsamplerState;

using DownsamplerOptionsProto =
    ::intrinsic_proto::data_logger::DownsamplerOptions;
using DownsamplerEventSourceStateProto =
    ::intrinsic_proto::data_logger::DownsamplerEventSourceState;
using DownsamplerStateProto = ::intrinsic_proto::data_logger::DownsamplerState;

using ::absl_testing::IsOkAndHolds;
using ::absl_testing::StatusIs;
using ::testing::HasSubstr;

// DownsamplerOptions.

TEST(DownsamplerOptions, RoundtripWorks) {
  DownsamplerOptions options{
      .sampling_interval_time = absl::Seconds(1),
      .sampling_interval_count = 10,
  };
  ASSERT_OK_AND_ASSIGN(DownsamplerOptionsProto proto, ToProto(options));
  EXPECT_THAT(FromProto(proto), IsOkAndHolds(options));

  DownsamplerOptions nullopt_options{
      .sampling_interval_time = std::nullopt,
      .sampling_interval_count = std::nullopt,
  };
  ASSERT_OK_AND_ASSIGN(DownsamplerOptionsProto nullopt_options_proto,
                       ToProto(nullopt_options));
  EXPECT_THAT(FromProto(nullopt_options_proto), IsOkAndHolds(nullopt_options));
}

TEST(DownsamplerOptions, ErrorsOnInvalidSamplingIntervalTime) {
  DownsamplerOptions invalid_options{.sampling_interval_time =
                                         absl::InfiniteDuration()};
  EXPECT_THAT(ToProto(invalid_options),
              StatusIs(absl::StatusCode::kInvalidArgument,
                       HasSubstr("invalid sampling_interval_time")));

  DownsamplerOptionsProto invalid_options_proto;
  invalid_options_proto.mutable_sampling_interval_time()->set_seconds(
      315576000001);
  EXPECT_THAT(FromProto(invalid_options_proto),
              StatusIs(absl::StatusCode::kInvalidArgument,
                       HasSubstr("invalid sampling_interval_time")));
}

// DownsamplerEventSourceState.

TEST(DownsamplerEventSourceState, RoundtripWorks) {
  DownsamplerEventSourceState state{
      .last_use_time = absl::FromUnixSeconds(123),
      .count_since_last_use = 42,
  };
  ASSERT_OK_AND_ASSIGN(DownsamplerEventSourceStateProto state_proto,
                       ToProto(state));
  EXPECT_THAT(FromProto(state_proto), IsOkAndHolds(state));
}

TEST(DownsamplerEventSourceState, ErrorsOnInvalidLastUseTime) {
  DownsamplerEventSourceState invalid_state{.last_use_time =
                                                absl::InfiniteFuture()};
  EXPECT_THAT(ToProto(invalid_state),
              StatusIs(absl::StatusCode::kInvalidArgument,
                       HasSubstr("invalid last_use_time")));

  DownsamplerEventSourceStateProto invalid_state_proto;
  invalid_state_proto.mutable_last_use_time()->set_seconds(253402300800);
  EXPECT_THAT(FromProto(invalid_state_proto),
              StatusIs(absl::StatusCode::kInvalidArgument,
                       HasSubstr("invalid last_use_time")));
}

// DownsamplerState.

TEST(DownsamplerState, RoundtripWorks) {
  DownsamplerState state{
      .event_source_states = {{"event_source_1",
                               {
                                   .last_use_time = absl::FromUnixSeconds(123),
                                   .count_since_last_use = 42,
                               }},
                              {"event_source_2",
                               {
                                   .last_use_time = absl::FromUnixSeconds(456),
                                   .count_since_last_use = 84,
                               }}},
  };
  ASSERT_OK_AND_ASSIGN(DownsamplerStateProto state_proto, ToProto(state));
  EXPECT_THAT(FromProto(state_proto), IsOkAndHolds(state));

  DownsamplerState empty_state{};
  ASSERT_OK_AND_ASSIGN(DownsamplerStateProto empty_state_proto,
                       ToProto(empty_state));
  EXPECT_THAT(FromProto(empty_state_proto), IsOkAndHolds(empty_state));
}

TEST(DownsamplerState, ErrorsOnInvalidLastUseTime) {
  DownsamplerState invalid_state{
      .event_source_states = {{"event_source_1",
                               {
                                   .last_use_time = absl::InfiniteFuture(),
                               }}},
  };
  EXPECT_THAT(ToProto(invalid_state),
              StatusIs(absl::StatusCode::kInvalidArgument,
                       HasSubstr("invalid last_use_time")));

  DownsamplerStateProto invalid_state_proto;
  (*invalid_state_proto.mutable_event_source_states())["event_source_1"]
      .mutable_last_use_time()
      ->set_seconds(253402300800);
  EXPECT_THAT(FromProto(invalid_state_proto),
              StatusIs(absl::StatusCode::kInvalidArgument,
                       HasSubstr("invalid last_use_time")));
}

}  // namespace
}  // namespace intrinsic

File: intrinsic/logging/utils/python/downsampler/proto_conversion_test.py
# Copyright 2023 Intrinsic Innovation LLC

"""Tests for the downsampler pybind module."""

import datetime

from absl.testing import absltest

from intrinsic.logging.proto import downsampler_pb2
from intrinsic.logging.utils.python.downsampler.downsampler import DownsamplerEventSourceState
from intrinsic.logging.utils.python.downsampler.downsampler import DownsamplerOptions
from intrinsic.logging.utils.python.downsampler.downsampler import DownsamplerState

from pybind11_abseil import status  # isort: skip

DownsamplerStateProto = downsampler_pb2.DownsamplerState
DownsamplerOptionsProto = downsampler_pb2.DownsamplerOptions


class DownsamplerOptionsTest(absltest.TestCase):

  def test_roundtrip_works(self):
    options = DownsamplerOptions(
        sampling_interval_time=datetime.timedelta(seconds=1),
        sampling_interval_count=10,
    )
    options_proto = DownsamplerOptions.to_proto(options)
    self.assertEqual(DownsamplerOptions.from_proto(options_proto), options)

    nullopt_options = DownsamplerOptions(
        sampling_interval_time=None, sampling_interval_count=None
    )
    nullopt_options_proto = DownsamplerOptions.to_proto(nullopt_options)
    self.assertEqual(
        DownsamplerOptions.from_proto(nullopt_options_proto), nullopt_options
    )

  def test_downsampler_options_errors_on_invalid_sampling_interval_time(self):
    options = DownsamplerOptions(sampling_interval_time=datetime.timedelta.max)
    with self.assertRaisesRegex(
        (status.StatusNotOk, RuntimeError), "invalid sampling_interval_time"
    ):
      DownsamplerOptions.to_proto(options)

    options_proto = DownsamplerOptionsProto(
        sampling_interval_time=datetime.timedelta(seconds=315576000001)
    )
    with self.assertRaisesRegex(
        (status.StatusNotOk, RuntimeError), "invalid sampling_interval_time"
    ):
      DownsamplerOptions.from_proto(options_proto)


class DownsamplerEventSourceStateTest(absltest.TestCase):

  def test_downsampler_event_source_state_roundtrip_works(self):
    state = DownsamplerEventSourceState(
        last_use_time=datetime.datetime.fromtimestamp(
            123, tz=datetime.timezone.utc
        ),
        count_since_last_use=42,
    )
    state_proto = DownsamplerEventSourceState.to_proto(state)
    self.assertEqual(DownsamplerEventSourceState.from_proto(state_proto), state)

  def test_downsampler_event_source_state_errors_on_invalid_last_use_time(self):
    invalid_state = DownsamplerEventSourceState(
        last_use_time=datetime.datetime.max, count_since_last_use=42
    )
    with self.assertRaisesRegex(
        (status.StatusNotOk, RuntimeError), "invalid last_use_time"
    ):
      DownsamplerEventSourceState.to_proto(invalid_state)

    # from_proto is covered by datetime checks.
    # i.e., datetime.fromtimestamp(253402300800, tz=datetime.timezone.utc errors


class DownsamplerStateTest(absltest.TestCase):

  def test_downsampler_state_roundtrip_works(self):
    state = DownsamplerState(
        event_source_states={
            "event_source_1": DownsamplerEventSourceState(
                last_use_time=datetime.datetime.fromtimestamp(
                    123, tz=datetime.timezone.utc
                ),
                count_since_last_use=42,
            ),
            "event_source_2": DownsamplerEventSourceState(
                last_use_time=datetime.datetime.fromtimestamp(
                    456, tz=datetime.timezone.utc
                ),
                count_since_last_use=84,
            ),
        }
    )
    state_proto = DownsamplerState.to_proto(state)
    self.assertEqual(DownsamplerState.from_proto(state_proto), state)

    empty_state = DownsamplerState(event_source_states={})
    empty_state_proto = DownsamplerState.to_proto(empty_state)
    self.assertEqual(
        DownsamplerState.from_proto(empty_state_proto), empty_state
    )

  def test_downsampler_state_errors_on_invalid_last_use_time(self):
    invalid_state = DownsamplerState(
        event_source_states={
            "event_source_1": DownsamplerEventSourceState(
                last_use_time=datetime.datetime.max,
                count_since_last_use=42,
            ),
        }
    )
    with self.assertRaisesRegex(
        (status.StatusNotOk, RuntimeError), "invalid last_use_time"
    ):
      DownsamplerState.to_proto(invalid_state)

    invalid_state_proto = DownsamplerStateProto()
    invalid_state_proto.event_source_states[
        "event_source_1"
    ].last_use_time.seconds = 253402300800
    with self.assertRaisesRegex(
        (status.StatusNotOk, RuntimeError), "invalid last_use_time"
    ):
      DownsamplerState.from_proto(invalid_state_proto)


if __name__ == "__main__":
  absltest.main()
